File: workout_service/src/main/java/com/louisfiges/workout/validation/Validator.java
package com.louisfiges.workout.validation;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Component
public class Validator {

    private final int minStringLength;
    private final int maxStringLength;

    public Validator(
            @Value("${min.string.length}") int minStringLength,
            @Value("${max.string.length}") int maxStringLength
    ) {
        this.minStringLength = minStringLength;
        this.maxStringLength = maxStringLength;
    }

    public boolean isInvalidString(String str) {
        if (minStringLength < 0 || maxStringLength < minStringLength) {
            throw new IllegalStateException("Invalid validator configuration");
        }

        return !getStringValidationIssues(str).isEmpty();
    }

    public List<String> getStringValidationIssues(String str) {
        if (minStringLength < 0 || maxStringLength < minStringLength) {
            throw new IllegalStateException("Invalid validator configuration");
        }

        List<String> issues = new ArrayList<>();

        if (str.isEmpty()) {
            issues.add("required");
            return List.copyOf(issues);
        }

        if (!str.equals(str.trim())) {
            issues.add("leading_or_trailing_whitespace");
        }

        int length = str.length();
        if (length < minStringLength) {
            issues.add("too_short");
        } else if (length > maxStringLength) {
            issues.add("too_long");
        }

        String regex = "^[\\p{L}\\p{N}\\s\\-_'()&.,!?]+$";
        if (!str.matches(regex)) {
            issues.add("invalid_characters");
        }

        return List.copyOf(issues);
    }

    public String getStringValidationMessage(String label, String str) {
        List<String> issues = getStringValidationIssues(str);
        if (issues.isEmpty()) {
            return null;
        }

        if (issues.size() == 1 && "required".equals(issues.getFirst())) {
            return label + " is required";
        }

        List<String> clauses = issues.stream()
                .map(this::describeStringValidationIssue)
                .toList();
        if (clauses.size() == 1) {
            return label + " must " + clauses.getFirst();
        }

        if (clauses.size() == 2) {
            return label + " must " + clauses.getFirst() + " and " + clauses.getLast();
        }

        return label + " must " + String.join(", ", clauses.subList(0, clauses.size() - 1))
                + ", and " + clauses.getLast();
    }

    private String describeStringValidationIssue(String issue) {
        return switch (issue) {
            case "too_short" -> "be at least " + minStringLength + " characters long";
            case "too_long" -> "be at most " + maxStringLength + " characters long";
            case "leading_or_trailing_whitespace" -> "not start or end with whitespace";
            case "invalid_characters" -> "contain only letters, numbers, spaces, and common punctuation";
            default -> throw new IllegalArgumentException("Unknown string validation issue: " + issue);
        };
    }
}

File: frontend/src/utils/validator.ts
import config from "@/config/config"

const MIN_STRING_LENGTH = config.MIN_STRING_LENGTH
const MAX_STRING_LENGTH = config.MAX_STRING_LENGTH
const MIN_SETS = config.MIN_SETS
const MAX_SETS = config.MAX_SETS

type StringValidationIssue =
  | "too_short"
  | "too_long"
  | "leading_or_trailing_whitespace"
  | "invalid_characters"

function isInvalidNumber(num: number, min: number, max: number): boolean {
  return num < min || num > max || !Number.isFinite(num)
}

export function getStringValidationIssues(str: string): StringValidationIssue[] {
  if (str.length === 0) {
    return []
  }

  const issues: StringValidationIssue[] = []

  if (str !== str.trim()) {
    issues.push("leading_or_trailing_whitespace")
  }

  const length = str.length
  if (length < MIN_STRING_LENGTH) {
    issues.push("too_short")
  } else if (length > MAX_STRING_LENGTH) {
    issues.push("too_long")
  }

  // letters, numbers, spaces and common punctuation
  const regex = /^[\p{L}\p{N}\s\-_'()&.,!?]+$/u
  if (!regex.test(str)) {
    issues.push("invalid_characters")
  }

  return issues
}

function describeStringValidationIssue(issue: StringValidationIssue): string {
  switch (issue) {
    case "too_short":
      return `be at least ${MIN_STRING_LENGTH} characters long`
    case "too_long":
      return `be at most ${MAX_STRING_LENGTH} characters long`
    case "leading_or_trailing_whitespace":
      return "not start or end with whitespace"
    case "invalid_characters":
      return "contain only letters, numbers, spaces, and common punctuation"
  }
}

export function getStringValidationMessage(
  label: string,
  str: string,
): string | null {
  if (str.length === 0) {
    return `${label} is required`
  }

  const issues = getStringValidationIssues(str)
  if (issues.length === 0) {
    return null
  }

  const clauses = issues.map(describeStringValidationIssue)
  if (clauses.length === 1) {
    return `${label} must ${clauses[0]}`
  }

  if (clauses.length === 2) {
    return `${label} must ${clauses[0]} and ${clauses[1]}`
  }

  const lastClause = clauses[clauses.length - 1]
  return `${label} must ${clauses.slice(0, -1).join(", ")}, and ${lastClause}`
}

export function isInvalidString(str: string): boolean {
  return str.length === 0 || getStringValidationIssues(str).length > 0
}

export function isInvalidSets(sets: number): boolean {
  return isInvalidNumber(sets, MIN_SETS, MAX_SETS)
}
